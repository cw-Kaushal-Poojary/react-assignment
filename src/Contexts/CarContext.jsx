import { createContext, useState, useContext } from 'react';

// Create the context
const CarContext = createContext();

// Create the provider component
export const CarContextProvider = ({ children }) => {
  // Initialize state for filters and sorting
  const [filters, setFilters] = useState({ minBudget: 0, maxBudget: 50, fuel: [] });
  const [sortOption, setSortOption] = useState('');
  const [cars, setCars] = useState([]);

  return (
    <CarContext.Provider value={{ filters, setFilters, sortOption, setSortOption, cars, setCars }}>
      {children}
    </CarContext.Provider>
  );
};

// Custom hook to use the context
export const useCarContext = () => {
  return useContext(CarContext);
};
