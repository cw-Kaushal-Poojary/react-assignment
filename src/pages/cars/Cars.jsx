import styles from "./Cars.module.css";

// Import filter image
import filterIcon from "../../assets/filter.png";
import CarCard from "../../components/Car-Card/CarCard";
import { useEffect, useState } from "react";

const Cars = () => {
  const fuelOptions = [
    {
      id: 1,
      name: "Petrol",
    },
    {
      id: 2,
      name: "Diesel",
    },
    {
      id: 3,
      name: "CNG",
    },
    {
      id: 4,
      name: "LPG",
    },
    {
      id: 5,
      name: "Electric",
    },
    {
      id: 6,
      name: "Hybrid",
    },
  ];

  const [cars, setCars] = useState([]);

  useEffect(() => {
    fetchCars();
  }, []);

  const fetchCars = async () => {
    try {
      const response = await fetch(
        "https://dummyjson.com/c/3151-8062-475b-8f20"
      );
      const data = await response.json();
      setCars(data.stocks);
    } catch (error) {
      console.error(error);
    }
  };

  console.log(cars);
  return (
    <div className={styles["cars-container"]}>
      <div className={styles["filter-container"]}>
        <div className={styles["filter-header"]}>
          <div className={styles["filter-text"]}>
            <span className={styles["filter-icon"]}>
              <img src={filterIcon} alt="" height={20} width={20} />
            </span>
            Filters
          </div>
          <button className={styles["clear-button"]}>Clear All</button>
        </div>
        <div className={styles["filter-section"]}>
          <div className={styles["filter-title"]}>Budget (Lakh)</div>
          <div className={styles["filter-budget"]}>
            <input type="number" placeholder="0" className={styles["input"]} />
            <span className={styles["dash"]}>-</span>
            <input type="number" placeholder="21" className={styles["input"]} />
          </div>
        </div>
        <div className={styles["filter-section"]}>
          <div className={styles["filter-title"]}>Fuel</div>
          <ul className={styles["fuel-options"]}>
            {fuelOptions?.map((fuelOption) => (
              <li key={fuelOption.id}>
                <input type="checkbox" /> {fuelOption.name}
              </li>
            ))}
          </ul>
        </div>
      </div>
      <div className={styles["sort-car-wrap"]}>
        <div className={styles["sort-cars"]}>
          <span>Sort By: </span>
          <select name="cars" id="cars">
            <option value="volvo">Best Match</option>
            <option value="saab">Price - Low to High</option>
            <option value="mercedes">Price - High to Low</option>
            <option value="audi">Year - Newest to Oldest</option>
            <option value="audi">Km - Low to High</option>
            <option value="audi">Km - High to Low</option>
          </select>
        </div>
        <div className={styles["card-container"]}>
          {cars?.map((car, index) => (
            <CarCard key={index} car={car} />
          ))}
        </div>
      </div>
    </div>
  );
};

export default Cars;
