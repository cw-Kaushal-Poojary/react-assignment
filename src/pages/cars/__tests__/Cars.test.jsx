// MODULES //
import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";

// COMPONENTS //
import Cars from "../Cars";

// CONTEXT //
import { useCarContext } from "../../../Contexts/CarContext";

// CONSTANTS //
import {
  dummyCars,
  dummyCarsAscending,
  dummyCarsDescending,
} from "../../../Constants";

// Mock CarContext
jest.mock("../../../Contexts/CarContext");

describe("Cars Component", () => {
  // Test Sorting
  let mockSetCars;

  beforeEach(() => {
    mockSetCars = jest.fn();
    fetchMock.mockResponseOnce(JSON.stringify({ stocks: dummyCars }));
    useCarContext.mockReturnValue({
      filters: { minBudget: 0, maxBudget: 50, fuel: [] },
      setFilters: jest.fn(),
      sortOption: "",
      setSortOption: jest.fn(),
      cars: dummyCars,
      setCars: mockSetCars,
    });
  });

  test("renders cars correctly", async () => {
    render(<Cars />);

    // Check if CarCard is rendered
    expect(await screen.getByText("Test Car")).toBeInTheDocument();
    expect(await screen.getByText("Test Car 2")).toBeInTheDocument();
  });

  test("sorts cars in ascending order", () => {
    render(<Cars />);

    const sortSelect = screen.getByTestId("sort-cars");
    fireEvent.change(sortSelect, { target: { value: "priceasc" } });

    expect(mockSetCars).toHaveBeenCalledWith(dummyCarsAscending);
  });

  test("sorts cars in descending order", () => {
    render(<Cars />);

    const sortSelect = screen.getByTestId("sort-cars");
    fireEvent.change(sortSelect, { target: { value: "pricedesc" } });

    expect(mockSetCars).toHaveBeenCalledWith(dummyCarsDescending);
  });
});
