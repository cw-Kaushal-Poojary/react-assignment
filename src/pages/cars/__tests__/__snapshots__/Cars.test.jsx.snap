// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cars Component matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <div>
        <div>
          <span>
            <img
              alt=""
              height="20"
              src="test-file-stub"
              width="20"
            />
          </span>
          Filters
        </div>
        <button>
          Clear All
        </button>
      </div>
      <div>
        <div>
          Budget (Lakh)
        </div>
        <div>
          <input
            data-testid="min-budget"
            placeholder="0"
            type="number"
            value="0"
          />
          <span>
            -
          </span>
          <input
            data-testid="max-budget"
            placeholder="50"
            type="number"
            value="50"
          />
        </div>
      </div>
      <div>
        <div>
          Fuel
        </div>
        <ul>
          <li>
            <input
              id="fuel-1"
              type="checkbox"
            />
            <label
              for="fuel-1"
            >
              Petrol
            </label>
          </li>
          <li>
            <input
              id="fuel-2"
              type="checkbox"
            />
            <label
              for="fuel-2"
            >
              Diesel
            </label>
          </li>
          <li>
            <input
              id="fuel-3"
              type="checkbox"
            />
            <label
              for="fuel-3"
            >
              CNG
            </label>
          </li>
          <li>
            <input
              id="fuel-4"
              type="checkbox"
            />
            <label
              for="fuel-4"
            >
              LPG
            </label>
          </li>
          <li>
            <input
              id="fuel-5"
              type="checkbox"
            />
            <label
              for="fuel-5"
            >
              Electric
            </label>
          </li>
          <li>
            <input
              id="fuel-6"
              type="checkbox"
            />
            <label
              for="fuel-6"
            >
              Hybrid
            </label>
          </li>
        </ul>
      </div>
    </div>
    <div>
      <div>
        <span>
          Sort By: 
        </span>
        <select
          data-testid="sort-cars"
          id="sort-cars"
          name="cars"
        >
          <option
            value=""
          >
            Choose an option
          </option>
          <option
            value="priceasc"
          >
            Price- Low to High
          </option>
          <option
            value="pricedesc"
          >
            Price- High to Low
          </option>
        </select>
      </div>
      <div>
        <div>
          <div>
            <img
              alt="car"
              height="100"
              src="test-image.jpg"
              width="100"
            />
            <div>
              <h3>
                Test Car
              </h3>
              <p>
                20000km | Petrol | Test Area, Test City
              </p>
              <p>
                Rs 15 Lakh
              </p>
              <div>
                <button>
                  Get Seller Details
                </button>
              </div>
            </div>
          </div>
        </div>
        <div>
          <div>
            <img
              alt="car"
              height="100"
              src="test-image-2.jpg"
              width="100"
            />
            <div>
              <h3>
                Test Car 2
              </h3>
              <p>
                30000km | Petrol | Test Area 2, Test City 2
              </p>
              <p>
                Rs 10 Lakh
              </p>
              <div>
                <button>
                  Get Seller Details
                </button>
              </div>
            </div>
          </div>
        </div>
        <div>
          <div>
            <img
              alt="car"
              height="100"
              src="test-image-3.jpg"
              width="100"
            />
            <div>
              <h3>
                Test Car 3
              </h3>
              <p>
                25000km | CNG | Test Area 3, Test City 3
              </p>
              <p>
                Rs 20 Lakh
              </p>
              <div>
                <button>
                  Get Seller Details
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
